vowpalwabbit/accumulate.cc

```diff
@@ -19,20 +19,26 @@ using namespace std;
 
 void add_float(float& c1, const float& c2) { c1 += c2; }
 
-void accumulate(vw& all, weight_parameters& weights, size_t offset)
-{ uint32_t length = 1 << all.num_bits; //This is size of gradient
+void accumulate(vw& all, parameters& weights, size_t offset)
+{ uint64_t length = 1 << all.num_bits; //This is size of gradient
   float* local_grad = new float[length];
 
-  weight_parameters::iterator iter = weights.begin();
-  for (uint32_t i = 0; iter != weights.end(); ++i, ++iter)
-    local_grad[i] = (&(*iter))[offset];
-
+  if (weights.sparse)
+    for (uint64_t i = 0; i < length; i++)
+      local_grad[i] = (&(weights.sparse_weights[i << weights.sparse_weights.stride_shift()]))[offset];
```

Comment:
sure that i << weights.sparse_weights.stride_shift() this gets optimized out?

again using template function this can be simplified